enforce docker restart before plugins setup

AlexeyShchukinSecurrency opened the pull request:
It fixes the issue when ansible successfully installs plugins but fails to run the last `restart docker` handler in the end.

haxorof commented:
LGMT! Thanks for you contribution!

haxorof commented:
Just notices that a start of Docker daemon happens just before you contribution of restarting the daemon. Can you describe when you get into the issue where it fails on the last restart? I see now it is not a clean install but happens during upgrade or reconfiguration? Is it easy to reproduce, in that case please write some simple steps.

AlexeyShchukinSecurrency commented:
### Environment:
- OS: Ubuntu 20.04
- ansible version: core 2.11.6

### Ansible configuration for docker_ce role
```yaml
docker_enable_audit: no

docker_plugins:
  - type: authz
    alias: openpolicyagent/opa-docker-authz-v2:0.8
    name: openpolicyagent/opa-docker-authz-v2:0.8
    args: opa-args="-policy-file /opa/policies/authz.rego"

docker_daemon_config:
  icc: false
  log-driver: journald
  log-level: info
  userns-remap: default
  live-restore: true
  userland-proxy: false
  no-new-privileges: true
  tls: true
  tlsverify: true
  tlscacert: /etc/docker/ssl/ca.crt
  tlscert: /etc/docker/ssl/server.crt
  tlskey: /etc/docker/ssl/server.key
  hosts:
    - "unix:///var/run/docker.sock"
    - "tcp://0.0.0.0:2376"
```


### Description
The error appears on **fresh install only**. Reproducibility is 100%.
If we run the same playbook with the same config, everything works as expected, meaning that there are no errors on the second try.

On the first run, `authz` plugin is installed successfully. If we stop at this point and execute `docker plugin ls`, it will show `authz` plugin is in place, but not in `enabled` state (as far as I remember).

```
TASK [haxorof.docker_ce : Install Docker plugins] ********************************************************************************************************************************
changed: [10.23.53.4] => (item={'type': 'authz', 'alias': 'openpolicyagent/opa-docker-authz-v2:0.8', 'name': 'openpolicyagent/opa-docker-authz-v2:0.8', 'args': 'opa-args="-policy-file /opa/policies/authz.rego"'})
```

However, at the end, ansible is unable to execute the last `restart docker` handler.

```
RUNNING HANDLER [restart docker] *************************************************************************************************************************************************
fatal: [10.23.53.4]: FAILED! => {"changed": false, "msg": "Unable to restart service docker: Job for docker.service failed because the control process exited with error code.\nSee \"systemctl status docker.service\" and \"journalctl -xe\" for details.\n"}
```

Right after that, there are many errors related to `authz` plugin in docker logs.

```bash
journalctl -xe -u docker
```

```
dockerd[137660]: time="2021-11-01T10:24:25.691181775Z" level=warning msg="Unable to locate plugin: openpolicyagent/opa-docker-authz-v2:0.8
dockerd[137660]: time="2021-11-01T10:24:26.692146845Z" level=warning msg="Unable to locate plugin: openpolicyagent/opa-docker-authz-v2:0.8
dockerd[137660]: time="2021-11-01T10:24:28.693397215Z" level=warning msg="Unable to locate plugin: openpolicyagent/opa-docker-authz-v2:0.8
dockerd[137660]: time="2021-11-01T10:24:32.696731739Z" level=warning msg="Unable to locate plugin: openpolicyagent/opa-docker-authz-v2:0.8
dockerd[137660]: failed to validate authorization plugin: plugin "openpolicyagent/opa-docker-authz-v2:0.8" not found
systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
```


### Purge docker
To reproduce the issue, it's enough to purge docker and run ansible once again.

```bash
apt purge docker-ce docker-ce-cli
```

```bash
rm -rf /etc/docker/ /etc/default/docker-envs /etc/systemd/system/docker.service.d/ /var/lib/docker/ /var/run/docker/
```

Unfortunately, I didn't have enough time to determine the root cause of such behavior. So, this fix should be considered as a workaround.

Please, let me know, if you need any additional details.

AlexeyShchukinSecurrency commented:
Thank you. I'm looking forward to the results.

haxorof commented:
After creating a manual regression test on this. Just want to inform you that I also manage to reproduce the problem you described with failing start of Docker daemon. Will look further into what is the root cause is.  Also at the same time I will correct the bug described in #142.

haxorof commented:
I have now run several tests with different Docker configurations to pin-point the issue. Reason the for failure in this case is related to `userns-remap: default` configuration. This relies on that you have configured your kernel correctly related to user namespaces. Removing that configuration will make it work and play successfully finishes and plugin is installed and enabled. I have since I implemented this in the role a [test related to Docker and CIS](https://github.com/haxorof/ansible-role-docker-ce/blob/master/tests/experimental/cis/test_cis.yml) (tested and works on CentOS 7 and AlmaLinux 8) and there you can see what I do to prepare the VM related to user namespaces for RHEL based OSes. Do not know what is required on Ubuntu to do similar to make it also work there.

@AlexeyShchukinSecurrency : Can you test and see if it also works for you if you remove `userns-remap: default` configuration?

AlexeyShchukinSecurrency commented:
Thank you for the update.

I've seen [test related to Docker and CIS](https://github.com/haxorof/ansible-role-docker-ce/blob/master/tests/experimental/cis/test_cis.yml). Kernel on my VM is configured according to the recommendations:
```bash
# cat /proc/cmdline
BOOT_IMAGE=/boot/vmlinuz-5.11.0-1020-azure root=PARTUUID=6b003b9b-0531-41bb-ab5e-b2491580c31f ro console=tty1 console=ttyS0 earlyprintk=ttyS0 apparmor=1 security=apparmor audit=1 audit_backlog_limit=8192 namespace.unpriv_enable=1 user_namespace.enable=1 panic=-1
```

Please, take a look at the list of `pre_tasks` I have, that are executed before the `docker_ce` role.

```yaml
    - name: PRE | DOCKER | TLS | Prepare required directories
      ansible.builtin.file:
        path: "{{ item.path }}"
        owner: "{{ item.owner }}"
        group: "{{ item.group }}"
        mode: "{{ item.mode }}"
        state: directory
      loop:
        - { path: '/etc/docker/ssl/', owner: 'root', group: 'root', mode: '0750' }
      when: docker_enabled
      become: true
      tags: [pre,docker]

    - name: PRE | DOCKER | TLS | Upload tls certs for docker daemon
      ansible.builtin.copy:
        dest: "{{ item.path }}"
        content: "{{ item.content }}"
        owner: root
        group: root
        mode: "{{ item.mode }}"
      loop:
        - { path: '/etc/docker/ssl/ca.crt', mode: '0440', content: "{{ docker_ca_crt }}" }
        - { path: '/etc/docker/ssl/server.crt', mode: '0440', content: "{{ docker_server_crt }}" }
        - { path: '/etc/docker/ssl/server.key', mode: '0400', content: "{{ docker_server_key }}" }
      no_log: true
      when: docker_enabled
      become: true
      tags: [pre,docker]

    - name: PRE | DOCKER | Enable user namespaces
      ansible.builtin.copy:
        dest: /etc/default/grub.d/99-user-namespaces.cfg
        content: |
          GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX namespace.unpriv_enable=1 user_namespace.enable=1"
        owner: root
        group: root
        mode: '0755'
      register: set_user_namespaces
      when: docker_enabled
      become: true
      tags: [pre,docker]

    - name: PRE | DOCKER | update grub on user namespaces changes
      ansible.builtin.command: update-grub
      when: set_user_namespaces.changed
      register: update_grub
      become: true
      tags: [pre,docker,grub]

    - name: PRE | DOCKER | reboot the system to apply changes
      ansible.builtin.reboot:
        reboot_timeout: 1200
      when: update_grub.changed
      become: true
      tags: [pre,docker,grub]

      # Related issue:
      #   https://github.com/haxorof/ansible-role-docker-ce/issues/131#issuecomment-722895050
      # The solution:
      #   https://github.com/haxorof/ansible-role-docker-ce/blob/master/tests/experimental/cis/test_cis.yml
    - name: PRE | DOCKER | Ensure /etc/docker/policies directory exists
      file:
        path: /etc/docker/policies
        state: directory
      when: docker_enabled
      become: true
      tags: [pre, docker, plugins, authz]

    - name: PRE | DOCKER | Create OPA policy file
      copy:
        dest: /etc/docker/policies/authz.rego
        content: |
          package docker.authz
          allow = true
      when: docker_enabled
      become: true
      tags: [pre, docker, plugins, authz]
```

Despite this, I'm still experiencing the issue.

haxorof commented:
Okay, just for sake of testing. Does it also work for you if you remove `userns-remap: default` from the `docker_daemon_config` configuration? Just want to confirm you see the same behavior as me. 🙂

AlexeyShchukinSecurrency commented:
I confirm everything works as expected without `userns-remap: default`.

haxorof commented:
I have now got time to investigate the root cause of this issue and restart as you suggested resolves the issue on Ubuntu and the reason why is because when Docker is installed on Ubuntu it is also started which is not the case for CentOS/AlmaLinux. I will write an issue on this and will commit fixes related to it before end of today.

I will close this PR but will reference it from the ticket I will write.

Thanks for reporting this and showing you configuration which show this problem.

AlexeyShchukinSecurrency commented:
I've just checked version `3.6.0` and everything works like a charm!
Thank you for your time and the awesome ansible role!
